### DISCUSSION TITLE
Recurse into subdirectories to find dot.yaml

### DISCUSSION THREAD

**Discussion by AdamYellen** [volllly/rotz#52] [Ideas]
Should Rotz recurse into all subdirectories? I'd like to organize some of my apps into subdirectories so my dotfiles folder isn't massive.

**Comment by volllly** [comment]
Hey, I like that idea 🙏

A few ways how to implement something like this come to mind.

### Automatically recurse empty folders

Just recurse empty folders until a `dot.yaml` file is found.

This does not need any new configuration.

This could lead to two dots with the same name though (if you `rotz install abc` and have the folders `def/abc/dot.yaml` and `ghi/abc/dot.yaml`). Could be solved by making dot names paths (`rotz install def/abc`).

### Add `workspace.yaml` file

Add a `workapace.yaml` file which works kind of like cargo workspaces.

You denote the folders you have there like so:

```yaml
workspaces:
  - somefolder
  - path: otherfolder
    name: othername
```

These folders then can contain a `dots.yaml` file and folders containing `dot.yaml` files like usual. Also they could contain an other `workspace.yaml` file for nested workspaces.

Add a global `--workspace` flag to rotz commands to select a workspace. In dependencies depend on dots from other workspaces like this `depends: [ workspace/nested:app ]`.

### Rename `dots.yaml` to `workspace.yaml`

Same as above but merge defaults and workspaces.

```yaml
workspaces:
  - somefolder
defaults:
  windows:
    installs: scoop install {{ name }}
```

> _Note: Support for `dots.yaml` files would be deprecated but not removed immediately._

---

I think I prefer something like the last option. Workspaces open the door for possible other features like importing another dotfile repo as a workspace. There's some configuration effort but I think it's fairly minimal

I'm interested to know what you think :)

**Reply by kecrily** [reply]
I love the first one. I think we already have workspace, each folder is a workspace. If we need to depend on other dot, then everything is fine relative to the root directory. Also, perhaps a `non-recursive` flag could be added.

```
rotz install something
```

Recursively install all the dots in the `something` folder


```
rotz install something --non-recursive
```

Install only `something/dots.yml`.

**Comment by AdamYellen** [comment]
Solution 1 and 2/3 are orthogonal. Both sound useful.

To solve the problem of duplicate names the user could always specify the full "path" if there is a conflict:
```
rotz install <workspace>/<dir>/<name>
```

**Comment by volllly** [comment]
Thanks @AdamYellen and @kecrily for your helpful input 🙏

I'm thinking of doing the following which is basically solution 1 😁

Recurse nested folders and treat dot names as paths. If the folders contain a defaults file merge it with the parent defaults and use it for child dots.

Dot names in the cli and the `depends` sections are glob patterns.
 
`rotz install devtools/*` installs all dots in the devtools folder. 

`rotz install **/vim` installs all dots named `vim` in any folder. 

`rotz install *` installs all dots in the root folder (like before).

`rotz install **` installs all dots in any folder.

`depends: [ ../package-managers/scoop ]` uses a path relative to the current dot.

I think this will work nicely without being to complex (Not like solution 2/3) and if needed we can still add a `workflow.yaml` file later on in a backwards compatible way.

**Reply by AdamYellen** [reply]
Sounds good.

**Comment by volllly** [comment]
Hey, the newest release [v0.7.0](https://github.com/volllly/rotz/releases/tag/v0.7.0) contains the functionality to nest dots 🎉

Let me know if something is not working as expected :)

**Reply by kecrily** [reply]
`rotz install` is fine, but `rotz link` throw `Warning: No dots found` when I run it in dotfiles root dir. Even when I specify dot manually, it says it can't find it.

**Reply by volllly** [reply]
Ah.. sorry about that.

Should be fixed in [v0.7.1](https://github.com/volllly/rotz/releases/tag/v0.7.1)
